sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java

```diff
@@ -450,36 +452,34 @@ private Object getValueFromJson(String propertyName, Method method) {
    */
   @SuppressWarnings({"unchecked", "rawtypes"})
   private Object getDefault(PipelineOptions proxy, Method method) {
+    if (method.getReturnType().equals(RuntimeValueProvider.class)) {
+      throw new RuntimeException(String.format(
+        "Method %s should not have return type "
+        + "RuntimeValueProvider, use ValueProvider instead.", method.getName()));
+    }
+    if (method.getReturnType().equals(StaticValueProvider.class)) {
+      throw new RuntimeException(String.format(
+        "Method %s should not have return type "
+        + "StaticValueProvider, use ValueProvider instead.", method.getName()));
+    }
```

Comment:
I'm actually a little confused here. It _looks like_ `options.getX()` where `X` is a `ValueProvider<T>` will *always* return a `RuntimeValueProvider<T>`. So why not only allow `RuntimeValueProvider<T>` there, rather than pretending it could be either? That would better convey "thou shalt not see this value until execution time".

For tests, maybe?